### PR TITLE
[Feature] Application deadline notification

### LINKED ISSUE #9640: ❗[Notification] Application deadline approaching
## ✨ Purpose

Remind a user that the deadline for a job application is approaching.

## ❗ Trigger

Send reminders 3 days before deadline.

This requires a cron job which runs every hour. It should look for currently open posters which close in the appropriate window (+/- 30 minutes?). 

## 👀 Audience

Notification should be sent to all users with a draft application for the closing poster.

### In-app string (English and French)

{{opportunity.title}} closes on {{opportunity_due_date_but_actually_the_day_before_EST}}. Continue your application.

FR: Le processus pour {{opportunity.title}} prend fin le 25 avril 2024. Continuer votre soumission.

### Email (English and French)

**Subject:** GC Digital Talent opportunity closes on April 25, 2024

_(already sent to translation by Marc.)_

**Body:**

You started an application to [{{opportunity.title}}](url) which closes in 3 days, at midnight on  {{opportunity_due_date_but_actually_the_day_before_EST}}.
If you’re still interested, we encourage you to [continue your application](url).

_(already sent to translation by Marc.)_

**Footer** (hopefully the same on all notification emails):
GC Digital Talent team

[Get in touch](url) with us to provide feedback or report a problem.

[GC Digital Talent](url) is a recruitment platform for digital jobs in the Government of Canada.

If you no longer want to receive these updates, you can [unsubscribe from notifications](url).
This is an automated message, so please do not reply.

_(already sent to translation by Marc.)_

### Formatting should look something like this... Ideally
image

### Links

Clicking the in-app notification should send you to continue the draft application. The "continue your application" text in the email should do the same.

In the email body, the "Opportunity Title" should be linked to the job advertisement for that process.

In the email footer "Get in touch" links to the support page https://talent.canada.ca/support .

In the email footer "you can unsubscribe from notifications" links to the new account settings page for managing notifications on the profile.

## 💾 Content

- This should be part of the **application updates** notification family (see #9555)

## ✅ Acceptance Criteria

- [ ] Email template exists in GC Notify
- [ ] Notification class exists in Laravel
- [ ] Notification is triggered by appropriate event and send to appropriate users (including NOT any users who are ignoring this notification's Notification Family) MAKE SURE TO USE TEST ENV
- [ ] Notification type added to schema
- [ ] Notification type rendered appropriately on frontend
- [ ] Tests? (How do we test the cron job?) DEV testing

## ⚙️ Deployment
- Ensure the cron job is running correctly on the server.

## 🛑 Blockers

Issues which must be completed before this one.

```[tasklist]
### Blocked By
- [ ] Should we make a separate ticket to get cron running on server?
```

### PR DESCRIPTION
**Pull request by petertgiles**
🤖 Resolves #9640 

## 👋 Introduction

This branch adds a notification to indicate if a pool that you have a draft application in will close soon.

## 🕵️ Details

There's a few moving parts here:
- A schedule loaded into Laravel Scheduler.
- An Artisan command that fires a notification to the right users.
- A notification that checks user preferences.
- A GCNotify email with templates in French and English.
- An new in-app notification.

### Locally

1. Make sure you're in the GCNotify team.
1. Add the GCNotify environment variables. (Ask me for them...)
2. Rebuild the app.
3. Create a pool that will close in three days.
4. Start (but don't submit) an application using your TBS email address to the pool.
5. `./artisan send-notifications:application-deadline-approaching`
6. Check for an email and an in-app notification.

### In Azure

1. Make sure you're in the GCNotify team.
2. Deploy the app to dev and do the following steps on the dev instance
3. Create a pool that will close in three days.
4. Start (but don't submit) an application using your TBS email address to the pool.
5. Wait until the next day
6. Check for an email and an in-app notification.

## 📸 Screenshot

image

## 🚚 Deployment

1. Ensure FEATURE_NOTIFICATIONS flag is still "false"
2. ~Set the GCNOTIFY_API_KEY variable to the appropriate one for each vertical (test, team, or prod)~
   - I set these.  Make sure IaC is released to at least Release-550. 
4. ~Set the GCNOTIFY_TEMPLATE_APPLICATION_DEADLINE_APPROACHING_EN variable the right ID (same for all verticals)~
    - I already did this but double-check, please
6. ~Set the GCNOTIFY_TEMPLATE_APPLICATION_DEADLINE_APPROACHING_FR variable the right ID (same for all verticals)~
    - I already did this but double-check, please

### CODE REVIEW

File: api/app/Enums/NotificationType.php
Comment (petertgiles) [inline api/app/Enums/NotificationType.php:1]: I think this enum is redundant.  I switched the type discriminator to use the built-in __typename graphql property.  The Laravel Notification system automatically creates a `type` column for itself to use, too.

File: api/config/notify.php [hunk 1]
```diff
@@ -28,6 +28,8 @@
         'test_sms' => env('GCNOTIFY_TEMPLATE_TEST_SMS'),
         'test_bulk_email' => env('GCNOTIFY_TEMPLATE_TEST_BULK_EMAIL'),
         'test_bulk_sms' => env('GCNOTIFY_TEMPLATE_TEST_BULK_SMS'),
+        'application_deadline_approaching_en' => env('GCNOTIFY_TEMPLATE_APPLICATION_DEADLINE_APPROACHING_EN'),
```
Comment (petertgiles) [inline api/config/notify.php:31]: If we're going to be making two of these for every new notification this is going to get out of hand in a hurry.  I don't know what the solution is, though.

File: api/app/Console/Kernel.php [hunk 2]
```diff
@@ -23,9 +24,15 @@ protected function schedule(Schedule $schedule)
         // clean up old user generated files every day at 2:00 AM
         $schedule->command(PruneUserGeneratedFiles::class)
             ->timezone('America/Toronto')
-            ->dailyAt('2:00')
+            ->dailyAt('1:00')
             ->withoutOverlapping()
             ->appendOutputTo('/tmp/laravel-prune-user-generated-files.log');
+
+        // queue up Application Deadline Approaching emails every day, close to the time the pool would close
+        $schedule->command(SendNotificationsApplicationDeadlineApproaching::class)
```
Comment (petertgiles) [inline api/app/Console/Kernel.php:32]: I ended up scheduling this daily instead of hourly like the issue described.  Daily is enough to reliably run this without missing any pools.  Also, if this runs hourly and is slightly off, you risk notifying a pool twice or not at all.

Comment (petertgiles) [reply api/app/Console/Kernel.php]: Gray confirmed that near the end-of-day in Newfoundland (when it is the same day across the country) is a good time to send the email.

File: api/app/Console/Kernel.php [hunk 3]
```diff
@@ -18,14 +20,30 @@ class Kernel extends ConsoleKernel
...
+        // Config utility does not seem to work in here!
+
         // $schedule->command(HardDeleteOldUsers::class)->dailyAt('08:00');
 
-        // clean up old user generated files every day at 2:00 AM
+        // clean up old user generated files every day at 1:00 AM
         $schedule->command(PruneUserGeneratedFiles::class)
             ->timezone('America/Toronto')
-            ->dailyAt('2:00')
+            ->dailyAt('1:00')
```
Comment (petertgiles) [inline api/app/Console/Kernel.php:31]: Unrelated fix.  2:00 is a terrible time to schedule something because of DST changes.  It never happens on "spring forward" days and happens twice on "fall back" days.

File: api/app/Enums/NotificationType.php [hunk 4]
```diff
@@ -10,6 +10,7 @@
 enum NotificationType
```
Comment (petertgiles) [inline api/app/Enums/NotificationType.php:10]: I think this enum is redundant. I switched the type discriminator to use the built-in __typename graphql property. The Laravel Notification system automatically creates a type column for itself to use, too.

File: api/app/Console/Kernel.php [hunk 5]
```diff
@@ -18,14 +20,30 @@ class Kernel extends ConsoleKernel
...
             ->withoutOverlapping()
             ->appendOutputTo('/tmp/laravel-prune-user-generated-files.log');
+
+        if (config('feature.notifications')) {
+
+            // queue up Application Deadline Approaching emails every day, close to the time the pool would close
+            $schedule->command(SendNotificationsApplicationDeadlineApproaching::class)
+                ->timezone('America/Toronto')
+                // 10 PM Eastern is the same day across the country, close to the end of the day in NL
+                ->dailyAt('22:00')
```
Comment (petertgiles) [inline api/app/Console/Kernel.php:41]: I ended up scheduling this daily instead of hourly like the issue described. Daily is enough to reliably run this without missing any pools. Also, if this runs hourly and is slightly off, you risk notifying a pool twice or not at all.

Gray confirmed that near the end-of-day in Newfoundland (when it is the same day across the country) is a good time to send the email.

File: api/.env.example [hunk 6]
```diff
@@ -90,6 +90,8 @@ GCNOTIFY_TEMPLATE_TEST_EMAIL=
 GCNOTIFY_TEMPLATE_TEST_SMS=
 GCNOTIFY_TEMPLATE_TEST_BULK_EMAIL=
 GCNOTIFY_TEMPLATE_TEST_BULK_SMS=
+GCNOTIFY_TEMPLATE_APPLICATION_DEADLINE_APPROACHING_EN=
+GCNOTIFY_TEMPLATE_APPLICATION_DEADLINE_APPROACHING_FR=
```
Comment (petertgiles) [inline api/.env.example:94]: If we're going to be making two of these for every new notification this is going to get out of hand in a hurry. I don't know what the solution is, though.

File: apps/web/src/hooks/useNotificationInfo.ts
Comment (petertgiles) [inline apps/web/src/hooks/useNotificationInfo.ts:1]: The refactoring got a little heavy in this file.  It seemed to me like the scaffolding that was here didn't work properly when a second notification type was added.

File: infrastructure/bin/post_deployment.sh [hunk 7]
```diff
@@ -71,7 +68,8 @@ fi
 add_section_block "$TRIPLE_BACK_TICK $CLEANED_STDOUT $TRIPLE_BACK_TICK"
 
 # Load Laravel Scheduler cron
-if echo "  *  *  *  *  * root    cd /home/site/wwwroot/api && php artisan schedule:run" >> /etc/crontab ; then
+# For extra debugging you can add `>> /tmp/run_laravel_scheduler.log 2>&1` to the end of the cron'd command
+if echo "  *  *  *  *  * root    . /etc/profile ; php /home/site/wwwroot/api/artisan schedule:run" >> /etc/crontab ; then
```
Comment (petertgiles) [inline infrastructure/bin/post_deployment.sh:72]: Oh man, this took me so long to figure out.  Turns out that processes launched from cron do not automatically have all the environment variables and they have to be added in manually.

File: api/.env.example (hunk 6 above)
Comment (brindasasi) [reply api/.env.example:94]: I checked out the GCNotify documentation. This is what how they expect us to use it.
Other ways I could think of is to make the body text also the placeholder/ message variable. 
Option 1: Define a message builder that can build different body of the messages based on what you want to send. Each message will re-use same template.
Option 2:  Re-use the template as much as we can. Like all application related messages one template and try to be as generic as possible and try to squeeze in deadline warning, status change updates to in single template. 
Anything like Password change could be diff template.

File: api/app/Console/Commands/SendNotificationsApplicationDeadlineApproaching.php [hunk 8]
```diff
@@ -0,0 +1,89 @@
...
+     */
+    public function handle()
+    {
+        $this->info('SendNotificationsApplicationDeadlineApproaching running at '.Carbon::now()->toDateTimeString());
+
+        $successCount = 0;
+        $failureCount = 0;
+
+        // server and database are running in UTC
+        $closingDayInPacific = Carbon::now('America/Vancouver')->addDays(3);
```
Comment (brindasasi) [inline api/app/Console/Commands/SendNotificationsApplicationDeadlineApproaching.php:40]: I feel like this 3 days needs to be dynamic. 
Although we decided on 3 days before closing date for now , this may change. 
It has to be a message variable in the template as well. 

You started an application to [((opportunity title))](https://notification.canada.ca/services/57b8682c-8400-40bd-a6bc-453b0c0fdd21/templates/mark%20class='placeholder'%3E((job%20advertisement%20link))%3C/mark) which closes in 3 days, at midnight on ((closing date)).

File: api/app/Console/Commands/SendNotificationsApplicationDeadlineApproaching.php [hunk 9]
```diff
@@ -0,0 +1,89 @@
...
+                    $pool->classification->displayName.': '.$pool->name['en'],
+                    $pool->classification->displayName.' : '.$pool->name['fr'],
+                    $pool->id,
+                    $poolCandidate->id,
+                );
+                try {
+                    $poolCandidate->user->notify($notification);
+                    $successCount++;
+                } catch (Throwable $e) {
+                    $failureCount++;
```
Comment (brindasasi) [inline api/app/Console/Commands/SendNotificationsApplicationDeadlineApproaching.php:77]: Is this a console command? Is this possible to get for whom the notifications failed to send?

File: api/app/Console/Kernel.php [hunk 10]
```diff
@@ -20,12 +22,25 @@ protected function schedule(Schedule $schedule)
...
+                ->appendOutputTo('/tmp/send-notifications-application-deadline-approaching.log');
```
Comment (brindasasi) [inline api/app/Console/Kernel.php:39]: oh ..it logs in here .. nvm then. Does it have failures too?

File: api/app/Enums/NotificationType.php (hunk 4 above)
Comment (brindasasi) [reply api/app/Enums/NotificationType.php:10]: I'm kinda confused by this. There is another NotificationType in the front end which has been used to define email or in-app notifications (https://github.com/GCTC-NTGC/gc-digital-talent/blob/1c1419cf935dd13158248b3f90c81acd9e99a225/apps/web/src/pages/Profile/AccountSettings/types.ts#L8) . If you don't need this can we get rid of it?

File: api/app/Models/Classification.php [hunk 11]
```diff
@@ -39,4 +40,15 @@ public function genericJobTitles(): HasMany
     {
         return $this->hasMany(GenericJobTitle::class);
     }
+
+    /**
+     * Get the classification display name, e.g. IT-01
+     */
+    protected function displayName(): Attribute
```
Comment (brindasasi) [inline api/app/Models/Classification.php:47]: Has this been used somewhere?

File: api/app/Notifications/GcNotifyEmailChannel.php [hunk 12]
```diff
@@ -0,0 +1,31 @@
...
+
+        $response = Notify::sendEmail(
+            $message->emailAddress,
+            $message->templateId,
+            $message->messageVariables
+        );
+
+        if (! $response->successful()) {
+            $errorMessage = 'Notification failed to send on GcNotifyEmailChannel';
+            Log::error($errorMessage);
```
Comment (brindasasi) [inline api/app/Notifications/GcNotifyEmailChannel.php:26]: 🥳

File: api/config/notify.php [hunk 13]
```diff
@@ -40,4 +42,25 @@
...
+    /*
+    |--------------------------------------------------------------------------
+    | GC Notify Smoke Test
+    |--------------------------------------------------------------------------
+    |
+    | If you need to smoke test your integration with GC Notify on a regular
+    | basis, you should use the following smoke test email addresses and phone
+    | numbers.
+    |
+    | https://documentation.notification.canada.ca/en/testing.html#smoke-testing
```
Comment (brindasasi) [inline api/config/notify.php:55]: Awesome!

**Review by brindasasi — COMMENTED** [review]
This looks great to me so far. Nice work! 
I've got a draft app in dev and will check out the email notification tmo.

File: api/app/Models/Classification.php (hunk 11 above)
Comment (petertgiles) [reply api/app/Models/Classification.php:47]: Yeah, it's used to build the notification.  Gray said he wanted the longer form of the pool name in the notification.
https://github.com/GCTC-NTGC/gc-digital-talent/blob/1c1419cf935dd13158248b3f90c81acd9e99a225/api/app/Console/Commands/SendNotificationsApplicationDeadlineApproaching.php#L68

File: api/app/Console/Commands/SendNotificationsApplicationDeadlineApproaching.php (hunk 9 above)
Comment (petertgiles) [reply api/app/Console/Commands/SendNotificationsApplicationDeadlineApproaching.php:77]: Yeah, you can run this manually with `./artisan send-notifications:application-deadline-approaching`.  It will then run for the "current day in Pacific".

File: api/app/Console/Commands/SendNotificationsApplicationDeadlineApproaching.php (hunk 8 above)
Comment (petertgiles) [reply api/app/Console/Commands/SendNotificationsApplicationDeadlineApproaching.php:40]: I considered this then figured it was probably over-engineering.  Let's bring it up in standup today to see if there's value there.

File: api/app/Console/Kernel.php (hunk 10 above)
Comment (petertgiles) [reply api/app/Console/Kernel.php:39]: Good idea!  I've added that in [log errors](https://github.com/GCTC-NTGC/gc-digital-talent/pull/10230/commits/accccd441d8ecf28e68c671eba78ebdfd6ee2a6d).
image

File: api/app/Enums/NotificationType.php (hunk 4 above)
Comment (petertgiles) [reply api/app/Enums/NotificationType.php:10]: @esizer Are you OK if I remove this enum?  I think you added it just for type discrimination right?

Comment (esizer) [reply api/app/Enums/NotificationType.php:10]: Uhh, that was actually @yonikid15 who added it think :sweat_smile: I am fine with it being removed though :stuck_out_tongue:

Comment (petertgiles) [reply api/app/Enums/NotificationType.php:10]: Cool, thanks.  :grin: 

image

Comment (petertgiles) [reply api/app/Enums/NotificationType.php]: [remove enum](https://github.com/GCTC-NTGC/gc-digital-talent/pull/10230/commits/c19e52c3ddaf614dd67b5bf4aa7c3f8e9fa87cc7) and [remove enum](https://github.com/GCTC-NTGC/gc-digital-talent/pull/10230/commits/25cb849336ef93556f618e397cc688c158e1c7d6)

**Review by brindasasi — APPROVED** [review]
Yayy.. I received both English & French emails. French title starts with lowercase. Not sure if that's intended.
I had a pool closing in 2 days and did NOT receive the email for that as expected. 
Clicking on those links takes me to the application page in that language. 
Everything looks good! 

image

image

**Comment by petertgiles** [comment]
> French title starts with lowercase.

Yeah, it was lowercase in the supplied document.
image
He can update the templates in GCNotify if necessary.
